asynclocalstorage.md

```diff
@@ -0,0 +1,354 @@
+
+The `frame` is effectively an *immutable* map of `storage keys` to `store`
+value. We refer to this as the `frame`'s `storage context`. A `storage key`
+is associated with exactly one `AsyncLocalStorage` object instance.
+
+There is always a `root frame` whose `storage context` is always empty.
+
+Whenever a new value is associated with a given `storage key`, a new
+`frame` is created with the `value` associated with that `storage key`
+set in the new frame's storage context.
+
+Whenever a new `frame` is created, it inherits a *copy* of the current
+`frame`'s `storage context` before setting the new `value` associated
+with the storage key.
+
+In **pseudo-code**, this looks something like:  (this is **NOT** the API,
+this is a pesudo-code illustration of the abstract model)
+
+```js
+class AsyncLocalStorage {
+  #storageKey = {}; // the storage key is opaque and unique to this
+                    // ALS instance.
+  run(store, fn) {
+    return AsyncContextFrame.run(#storageKey, store, fn);
+  }
+  
+  getStore() {
+    return AsyncContextFrame::current().get(this.#storageKey);
+  }
+}
+
+class AsyncContextFrame {
+  static #currentFrame;
+  #storageContext = new Map();
+
+  static run(key, store, fn) {
+    const current = AsyncContextFrame::current();
+    const frame = new AsyncContextFrame(current, key, store);
+    return frame.run(fn);
+  }
+  
+  static current() { return AsyncContextFrame.#currentFrame; }
+  static exchange(frame) {
+    const prior = AsyncContextFrame.#currentFrame;
+    AsyncContextFrame.#currentFrame = frame;
+    return prior;
+  }
+  
+  constructor(current, key, store) {
+    current.#storageContext.propagateTo(this.#storageContext);
+    this.#storageContext.set(key, store);
+  }
+  
+  run(fn) {
+    // enter this frame...
+    const prior = AsyncContextFrame::exchange(this);
+    try {
+      return fn();
+    } finally {
+      AsyncContextFrame:exchange(prior);
+    }
+  }
+  
+  get(key) {
+    return this.#storageContext.get(key);
+  }
+}
+```
+
+Any async resource that is created needs only to grab the current `frame`
+and run code within its scope to access the correct `storage context`.
+
+Note that key to this model is that the `storage context` is *immutable* 
+once created. Mutations to the context using `asyncLocalStorage.run()`
+use copy-on-write semantics, causing a new `frame` to be created and
+entered. This is a key performance optimization over the current model
+implemented by Node.js.
+
+## Class: AsyncLocalStorage
+
+The following is the limited subset of the `AsyncLocalStorage` API:
+
+```js
+class AsyncLocalStorage {
+  constructor();
+  
+  // Creates a new frame, propagating the storage context of
+  // the current frame to the new frame, then setting the store
+  // as the value associated with this ALS instance. The function
+  // is then run within the context of the newly created frame.
+  // The return value of calling fn() is returned by run().
+  run(store, fn) : any;
```

Comment:
Is the rest argument list removed intentionally?

```suggestion
  run(store, fn, ...args) : any;
```

Reply:
Sorry, no, just an oversight!